File: 0692-top-k-frequent-words/0692-top-k-frequent-words.py
class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        freq = defaultdict(int)
        for word in words:
            freq[word] += 1

        bucket = [[] for _ in range(len(words) + 1)]

        for word, count in freq.items():
            bucket[count].append(word)

        ans = []

        for i in range(len(words), 0, -1):
            if not bucket[i]:
                continue
            bucket[i].sort()

            for word in bucket[i]:
                ans.append(word)
                if len(ans) == k:
                    return ans

        return ans

File: 0692-top-k-frequent-words/0692-top-k-frequent-words.java
class Solution {
    public List<String> topKFrequent(String[] words, int k) {
        HashMap<String, Integer> map = new HashMap<>();

        for (String word : words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
        }

        List<String>[] bucket = new List[words.length + 1];
        for (Map.Entry<String, Integer> entry : map.entrySet()) {
            int frequency = entry.getValue();

            if (bucket[frequency] == null) {
                bucket[frequency] = new ArrayList<>();
            }

            bucket[frequency].add(entry.getKey());
        }

        List<String> answer = new ArrayList<>();
        for (int freq = words.length; freq >= 1; freq--) {
            if (bucket[freq] == null) {
                continue;
            }
            Collections.sort(bucket[freq]);

            for (String word : bucket[freq]) {
                answer.add(word);
                if (answer.size() == k) {
                    return answer;
                }
            }
        }

        return answer;
    }
}